addons/web/static/tests/fields/basic_fields_tests.js

```diff
@@ -2682,7 +2682,11 @@ QUnit.module('basic_fields', {
         // set an invalid date
         var $input = form.$('.o_field_widget[name=date] input');
         $input.val('mmmh').trigger('change');
-        assert.strictEqual($input.text(), "", "The date field should be empty");
+        // save the change
+        form.$buttons.find('.o_form_button_save').click();
+        assert.strictEqual($input.val(), "mmmh", "The wrong value should be kept");
+        var dateField = _.find(form.renderer.allFieldWidgets)[0];
+        assert.strictEqual(dateField.isValid(), false, "The date field is not valid");
```

Comment:
This is a implementation detail. I'd only check what is observable, that is, the input content and the fact that the field input is correctly marked as invalid (it should have a classname o_invalid or something). Also, a dialog should be displayed. You could also check that the form is still in edit mode (o_form_edit).